omniledger/service/struct.go

```diff
@@ -175,3 +179,86 @@ type Data struct {
 	Timestamp    int64
 	Roster       *onet.Roster
 }
+
+// sortTransactions needs to marshal transactions, if it fails to do so,
+// it returns an error and leaves the slice unchange.
+func sortTransactions(ts []Transaction) error {
+	bs := make([][]byte, len(ts))
+	sortedTs := make([]*Transaction, len(ts))
+	var err error
+	var ok bool
+	for i := range ts {
+		bs[i], err = network.Marshal(&ts[i])
+		if err != nil {
+			return err
+		}
+	}
+	// An alternative to XOR-ing the transactions would have been to
+	// concatenate them and hash the result. However, if we generate the salt
+	// as the hash of the concatenation of the transactions, we have to
+	// concatenate them in a specific order to be deterministic.
+	// This means we would have to sort them, just to get the salt.
+	// In order to avoid this, we XOR them.
+	salt := xorTransactions(bs)
+	sortWithSalt(bs, salt)
+	for i := range bs {
+		_, tmp, err := network.Unmarshal(bs[i], cothority.Suite)
+		if err != nil {
+			return err
+		}
+		sortedTs[i], ok = tmp.(*Transaction)
+		if !ok {
+			return errors.New("Data of wrong type")
+		}
+	}
+	for i := range sortedTs {
+		ts[i] = *sortedTs[i]
+	}
+	return nil
+}
+
+func sortWithSalt(ts [][]byte, salt []byte) {
+	less := func(i, j int) bool {
+		h := sha256.New()
+		h.Write(append(salt, ts[i]...))
+		h1 := h.Sum(nil)
+		h.Reset()
+		h.Write(append(salt, ts[j]...))
+		h2 := h.Sum(nil)
+		h.Reset()
```

Comment:
not needed here. As you're hashing only one element, you can even do:
```
h1 := sha256.Sum256(append(salt, ts[i]...))
h2 := sha256.Sum256(append(salt, ts[j]...))
return bytes.Compare(h1, h2)
```
It looks more concise and is clear enough.